flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -0,0 +1,300 @@
+		private final KinesisDeserializationSchema<T> deserializer;
+		private final HashMap<KinesisStreamShard, String> seqNoState;
+
+		private final KinesisProxy kinesisProxy;
+
+		private final KinesisDataFetcher ownerRef;
+
+		private final KinesisStreamShard assignedShard;
+
+		private String lastSequenceNum;
+		private String nextShardItr;
+
+		private volatile boolean running = true;
+
+		public ShardConsumerThread(KinesisDataFetcher ownerRef,
+								Properties props,
+								KinesisStreamShard assignedShard,
+								String lastSequenceNum,
+								SourceFunction.SourceContext<T> sourceContext,
+								KinesisDeserializationSchema<T> deserializer,
+								HashMap<KinesisStreamShard, String> seqNumState) {
+			this.ownerRef = checkNotNull(ownerRef);
+			this.assignedShard = checkNotNull(assignedShard);
+			this.lastSequenceNum = checkNotNull(lastSequenceNum);
+			this.sourceContext = checkNotNull(sourceContext);
+			this.deserializer = checkNotNull(deserializer);
+			this.seqNoState = checkNotNull(seqNumState);
+			this.kinesisProxy = new KinesisProxy(props);
+		}
+
+		@Override
+		public void run() {
+			try {
+				if (lastSequenceNum.equals(SentinelSequenceNumber.SENTINEL_LATEST_SEQUENCE_NUM.toString())) {
+					// if the shard is already closed, there will be no latest next record to get for this shard
+					if (assignedShard.isClosed()) {
+						nextShardItr = null;
+					} else {
+						nextShardItr = kinesisProxy.getShardIterator(assignedShard, ShardIteratorType.LATEST.toString(), null);
+					}
+				} else if (lastSequenceNum.equals(SentinelSequenceNumber.SENTINEL_EARLIEST_SEQUENCE_NUM.toString())) {
+					nextShardItr = kinesisProxy.getShardIterator(assignedShard, ShardIteratorType.TRIM_HORIZON.toString(), null);
+				} else if (lastSequenceNum.equals(SentinelSequenceNumber.SENTINEL_SHARD_ENDING_SEQUENCE_NUM.toString())) {
+					nextShardItr = null;
+				} else {
+					nextShardItr = kinesisProxy.getShardIterator(assignedShard, ShardIteratorType.AFTER_SEQUENCE_NUMBER.toString(), lastSequenceNum);
+				}
+
+				long lastNextShardItrUpdateMillis = System.currentTimeMillis();
+				boolean noRecordsOnLastFetch = false;
+				while(running) {
+					if (nextShardItr == null) {
+						lastSequenceNum = SentinelSequenceNumber.SENTINEL_SHARD_ENDING_SEQUENCE_NUM.toString();
+
+						synchronized (sourceContext.getCheckpointLock()) {
+							seqNoState.put(assignedShard, lastSequenceNum);
+						}
+
+						break;
+					} else {
+						if (noRecordsOnLastFetch) {
+							if (System.currentTimeMillis() - lastNextShardItrUpdateMillis >= 290000) {
+								nextShardItr = kinesisProxy.getShardIterator(assignedShard, ShardIteratorType.AFTER_SEQUENCE_NUMBER.toString(), lastSequenceNum);
+								lastNextShardItrUpdateMillis = System.currentTimeMillis();
+							}
+						}
+
+						GetRecordsResult getRecordsResult = kinesisProxy.getRecords(nextShardItr, 100);
+
+						final List<Record> fetchedRecords = getRecordsResult.getRecords();
```

Comment:
The records returned here might be aggregated: http://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-consumer-deaggregation.html
So we need to add code to deaggregate them here.